### PR TITLE
add stats tagger APIs and connection stats.

### PR DESCRIPTION
**Pull request by menghanl**

### CODE REVIEW

File: stats/handlers.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+// For RPC stats handling,
+//  - On server side, the context used in RPC stats handler for all RPCs using this
+// connection will be derived from the context returned.
+//  - On client side, the context is not derived from the context returned.
+func RegisterConnCtxTagger(t func(context.Context, *ConnContextTagInfo) context.Context) {
+	connTagger = t
+}
+
+// TagRPCCtx calls user registered RPC context tagger.
+func TagRPCCtx(ctx context.Context, info *RPCContextTagInfo) context.Context {
```
Comment (iamqizhao) [inline stats/handlers.go]: RPCTagger?

File: stats/handlers.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+// ConnContextTagInfo defines the relevant information needed by connection context tagger.
+type ConnContextTagInfo struct {
+	// RemoteAddr is the remote address of the corresponding connection.
+	RemoteAddr net.Addr
+	// LocalAddr is the local address of the corresponding connection.
+	LocalAddr net.Addr
+}
+
+// RPCContextTagInfo defines the relevant information needed by RPC context tagger.
+type RPCContextTagInfo struct {
```
Comment (iamqizhao) [inline stats/handlers.go]: I would prefer RPCTagInfo actually.

File: stats/handlers.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import (
+	"net"
+	"sync/atomic"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/grpclog"
+)
+
+// ConnContextTagInfo defines the relevant information needed by connection context tagger.
+type ConnContextTagInfo struct {
```
Comment (iamqizhao) [inline stats/handlers.go]: I prefer ConnTagInfo.

File: stats/handlers.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+// RegisterConnHandler registers the user handler function for conn stats.
+// If another handler was registered before, this new handler will overwrite the old one.
+// This handler function will be called to process the conn stats.
+func RegisterConnHandler(f func(context.Context, ConnStats)) {
+	connhandler = f
+}
+
+// TagConnCtx calls user registered connection context tagger.
+func TagConnCtx(ctx context.Context, info *ConnContextTagInfo) context.Context {
```
Comment (iamqizhao) [inline stats/handlers.go]: ConnTagger?

File: stats/handlers.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+		return ctx
+	}
+	return rpcTagger(ctx, info)
+}
+
+// RegisterRPCCtxTagger registers the user RPC context tagger function.
+// The RPC context tagger can attach some information to the given context.
+// The context used in stats handler for this RPC will be derived from the
+// context returned.
+func RegisterRPCCtxTagger(t func(context.Context, *RPCContextTagInfo) context.Context) {
```
Comment (iamqizhao) [inline stats/handlers.go]: RegisterRPCTagger?

File: stats/stats.go [hunk 6]
```diff
@@ -181,39 +191,33 @@ type End struct {
...
+type ConnStats interface {
+	isConnStats()
+	// IsClient returns true if this ConnStats is from client side.
+	IsClient() bool
 }
 
-// Handle processes the stats using the call back function registered by user.
-func Handle(ctx context.Context, s RPCStats) {
-	handler(ctx, s)
+// ConnBegin contains stats when an RPC begins.
```
Comment (iamqizhao) [inline stats/stats.go]: ConnBegin has the stats of a connection when it is established.

File: stats/stats.go [hunk 7]
```diff
@@ -181,39 +191,33 @@ type End struct {
...
-	}
-	atomic.StoreInt32(on, 1)
-}
+func (s *ConnBegin) isConnStats() {}
 
-// Stop stops the stats collection and processing.
-// Stop does not unregister handler.
-func Stop() {
-	atomic.StoreInt32(on, 0)
+// ConnEnd contains stats when an RPC ends.
```
Comment (iamqizhao) [inline stats/stats.go]: ConnEnd has the states of a connection when it ends.

File: transport/http2_client.go [hunk 8]
```diff
@@ -181,6 +182,7 @@ func newHTTP2Client(ctx context.Context, addr TargetInfo, opts ConnectOptions) (
 	}
 	var buf bytes.Buffer
 	t := &http2Client{
+		ctx:        context.Background(),
```
Comment (iamqizhao) [inline transport/http2_client.go]: use the one passed by NewClientTransport instead?

File: stats/handlers.go (hunk 4 above)
Comment (menghanl) [inline stats/handlers.go]: Changed it to TagConn()

File: stats/handlers.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+// RPCTagInfo defines the relevant information needed by RPC context tagger.
+type RPCTagInfo struct {
+	// FullMethodName is the string of gRPC method (in the format of /package.service/method).
+	FullMethodName string
+}
+
+var (
+	on          = new(int32)
+	handler     func(context.Context, RPCStats)
```
Comment (iamqizhao) [inline stats/handlers.go]: rpcHandler

File: stats/handlers.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	connHandler func(context.Context, ConnStats)
+	connTagger  func(context.Context, *ConnTagInfo) context.Context
+	rpcTagger   func(context.Context, *RPCTagInfo) context.Context
+)
+
+// Handle processes the RPC stats using the call back function registered by user.
+func Handle(ctx context.Context, s RPCStats) {
```
Comment (iamqizhao) [inline stats/handlers.go]: HandleRPC

File: stats/handlers.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	if handler == nil {
+		return
+	}
+	handler(ctx, s)
+}
+
+// RegisterHandler registers the user handler function for rpc stats.
+// If another handler was registered before, this new handler will overwrite the old one.
+// This handler function will be called to process the rpc stats.
+func RegisterHandler(f func(context.Context, RPCStats)) {
```
Comment (iamqizhao) [inline stats/handlers.go]: RegisterRPCHandler

File: stats/handlers.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	atomic.StoreInt32(on, 1)
+}
+
+// Stop stops the stats collection and processing.
+// Stop does not unregister handlers.
+func Stop() {
+	atomic.StoreInt32(on, 0)
+}
+
+// On indicates whether stats is started.
```
Comment (iamqizhao) [inline stats/handlers.go]: "... whether the stats collection and processing is on.

File: stats/handlers.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+// RegisterRPCTagger registers the user RPC context tagger function.
+// The RPC context tagger can attach some information to the given context.
+// The context used in stats rpcHandler for this RPC will be derived from the
+// context returned.
+func RegisterRPCTagger(t func(context.Context, *RPCTagInfo) context.Context) {
+	rpcTagger = t
+}
+
+// Start starts the stats collection and reporting if there is a registered stats handle.
```
Comment (iamqizhao) [inline stats/handlers.go]: s/reporting/processing

File: stats/handlers.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+// connection will be derived from the context returned.
+// For RPC stats handling,
+//  - On server side, the context used in rpcHandler for all RPCs using this
+// connection will be derived from the context returned.
+//  - On client side, the context is not derived from the context returned.
+func RegisterConnTagger(t func(context.Context, *ConnTagInfo) context.Context) {
+	connTagger = t
+}
+
+// TagRPC calls user registered RPC context tagger.
```
Comment (iamqizhao) [inline stats/handlers.go]: s/user/the user

File: stats/handlers.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+	// RemoteAddr is the remote address of the corresponding connection.
+	RemoteAddr net.Addr
+	// LocalAddr is the local address of the corresponding connection.
+	LocalAddr net.Addr
+	// TODO add QOS related fields.
+}
+
+// RPCTagInfo defines the relevant information needed by RPC context tagger.
+type RPCTagInfo struct {
+	// FullMethodName is the string of gRPC method (in the format of /package.service/method).
```
Comment (iamqizhao) [inline stats/handlers.go]: FullMethodName is the RPC method in the format of /package.service/method.

File: stats/handlers.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+var (
+	on          = new(int32)
+	rpcHandler  func(context.Context, RPCStats)
+	connHandler func(context.Context, ConnStats)
+	connTagger  func(context.Context, *ConnTagInfo) context.Context
+	rpcTagger   func(context.Context, *RPCTagInfo) context.Context
+)
+
+// HandleRPC processes the RPC stats using the call back function registered by user.
```
Comment (iamqizhao) [inline stats/handlers.go]: ... using the rpc handler registered by the user.

File: stats/handlers.go [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+func HandleRPC(ctx context.Context, s RPCStats) {
+	if rpcHandler == nil {
+		return
+	}
+	rpcHandler(ctx, s)
+}
+
+// RegisterRPCHandler registers the user handler function for rpc stats.
```
Comment (iamqizhao) [inline stats/handlers.go]: ... for RPC stats processing.

File: stats/handlers.go [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+// If another handler was registered before, this new handler will overwrite the old one.
```
Comment (iamqizhao) [inline stats/handlers.go]: It should be called only once. The later call will overwrite the former value if it is called multiple times.

File: stats/handlers.go [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+// HandleConn processes the stats using the call back function registered by user.
+func HandleConn(ctx context.Context, s ConnStats) {
+	if connHandler == nil {
+		return
+	}
+	connHandler(ctx, s)
+}
+
+// RegisterConnHandler registers the user handler function for conn stats.
+// If another handler was registered before, this new handler will overwrite the old one.
```
Comment (iamqizhao) [inline stats/handlers.go]: ditto

File: stats/handlers.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+	return connTagger(ctx, info)
+}
+
+// RegisterConnTagger registers the user connection context tagger function.
+// The connection context tagger can attach some information to the given context.
+// The returned context will be used for stats handling.
+// For conn stats handling, the context used in connHandler for this
+// connection will be derived from the context returned.
+// For RPC stats handling,
+//  - On server side, the context used in rpcHandler for all RPCs using this
```
Comment (iamqizhao) [inline stats/handlers.go]: s/using/on

File: stats/handlers.go [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+func Start() {
+	if rpcHandler == nil && connHandler == nil {
+		grpclog.Println("rpcHandler and connHandler are both nil when starting stats. Stats is not started")
+		return
+	}
+	atomic.StoreInt32(on, 1)
+}
+
+// Stop stops the stats collection and processing.
+// Stop does not unregister handlers.
```
Comment (iamqizhao) [inline stats/handlers.go]: the handlers.

**Pull request merged**